wallet/src/key_chain/leaf_key_chain.rs

```diff
@@ -0,0 +1,517 @@
+            Ok(())
+        }
+    }
+
+    /// Derives a key or gets it from the precomputed key pool.
+    fn derive_key(&self, key_index: ChildNumber) -> KeyChainResult<ExtendedPublicKey> {
+        // Get the public key from the key pool if available
+        if let Some(pub_key) = self.derived_public_keys.get(&key_index) {
+            return Ok(pub_key.clone());
+        }
+
+        // Create the new key path
+        let key_path = {
+            let mut path = self.parent_pubkey.get_derivation_path().clone().into_vec();
+            path.push(key_index);
+            path.try_into()?
+        };
+
+        // Derive the key
+        Ok(self.parent_pubkey.clone().derive_absolute_path(&key_path)?)
+    }
+
+    /// Derives and adds a key to his key chain. This does not affect the last used and issued state
+    fn derive_and_add_key<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        key_index: ChildNumber,
+    ) -> KeyChainResult<ExtendedPublicKey> {
+        // Get the public key from the key pool if available
+        if let Some(pub_key) = self.derived_public_keys.get(&key_index) {
+            return Ok(pub_key.clone());
+        }
+
+        // Derive the new extended public key
+        let derived_key = self.derive_key(key_index)?;
+        // Just the public key
+        let public_key = derived_key.clone().into_public_key();
+        // Calculate public key hash
+        let public_key_hash = PublicKeyHash::from(&public_key);
+        // Calculate the address
+        let address = Address::from_public_key_hash(&self.chain_config, &public_key_hash)?;
+        // Calculate account derivation path id
+        let account_path_id = AccountDerivationPathId::new(
+            self.account_id.clone(),
+            derived_key.get_derivation_path().clone(),
+        );
+
+        // Save issued key and address
+        db_tx.set_public_key(&account_path_id, &derived_key)?;
+        db_tx.set_address(&account_path_id, &address)?;
+
+        // Add key and address the maps
+        self.derived_public_keys.insert(key_index, derived_key.clone());
+        self.addresses.insert(key_index, address);
+        self.public_key_to_index.insert(public_key, key_index);
+        self.public_key_hash_to_index.insert(public_key_hash, key_index);
+
+        Ok(derived_key)
+    }
+
+    /// Get the last derived key index
+    pub fn get_last_derived_index(&self) -> Option<ChildNumber> {
+        self.derived_public_keys.last_key_value().map(|(k, _)| *k)
+    }
+
+    /// Derive up `lookahead_size` keys starting from the last used index. If the gap from the last
+    /// used key to the last derived key is already `lookahead_size`, this method has no effect
+    pub fn top_up<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        lookahead_size: u32,
+    ) -> KeyChainResult<()> {
+        // Find how many keys to derive
+        let (starting_index, up_to_index) = match self.get_last_derived_index() {
```

Comment:
Maybe it would be better to rewrite it like this?
```
        let starting_index = match self.get_last_derived_index() {
            Some(last_derived_index) => last_derived_index.increment()?.get_index(),
            None => 0,
        };
        let up_to_index = match self.usage_state.get_last_used() {
            Some(last_used) => last_used.get_index() + lookahead_size + 1,
            None => lookahead_size,
        };
```